{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## In the   Accelerated_Intro_WilkinsonExams    folder\n",
    "\n",
    "## ISSUE THESE COMMADS \n",
    "\n",
    "\n",
    "###   git fetch\n",
    "###   git checkout FETCH_HEAD -- Exam\\ Week\\ 2.ipynb\n",
    "###   git checkout FETCH_HEAD -- Germplasm.tsv\n",
    "###   git checkout FETCH_HEAD -- LocusGene.tsv\n",
    "\n",
    "this will retrieve the exam material without destroying your existing notes\n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "## Start your mySQL Server from a terminal \n",
    "(if it isn't already running)\n",
    "\n",
    "<code>sudo docker start course-mysql</code>\n",
    "<pre>\n",
    "\n",
    "</pre>\n",
    "Don't forget that, if you use sqlMagic, you need to execute the connection lines in your Notebook!\n",
    "\n",
    "<pre>\n",
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "</pre>\n",
    "\n",
    "## Create a new Python3 Jupyter Notebook in your Exam Answers folder\n",
    "\n",
    "commit and push this Notebook when you are finished.\n",
    "\n",
    "You have **2 hours and 50 minutes to complete this exam.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Data Files\n",
    "Germplasm.tsv and LocusGene.tsv contain the datasets we need for the exam.\n",
    "\n",
    "Our objective is to create a database to contain the data in these files, insert the data into the database, then query the database in a variety of ways.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Line 0: Locus Locus => same sequence\n",
      "Line 1: AT1G01040 AT1G01040 => same sequence\n",
      "Line 2: AT1G01060 AT1G01060 => same sequence\n",
      "Line 3: AT1G01140 AT1G01140 => same sequence\n",
      "Line 4: AT1G01220 AT1G01220 => same sequence\n",
      "Line 5: AT2G03720 AT2G03720 => same sequence\n",
      "Line 6: AT2G03800 AT2G03800 => same sequence\n",
      "Line 7: AT2G04240 AT2G04240 => same sequence\n",
      "Line 8: AT2G05210 AT2G05210 => same sequence\n",
      "Line 9: AT3G02130 AT3G02130 => same sequence\n",
      "Line 10: AT3G02140 AT3G02140 => same sequence\n",
      "Line 11: AT3G02230 AT3G02230 => same sequence\n",
      "Line 12: AT3G02260 AT3G02260 => same sequence\n",
      "Line 13: AT3G02310 AT3G02310 => same sequence\n",
      "Line 14: AT3G02680 AT3G02680 => same sequence\n",
      "Line 15: AT3G02850 AT3G02850 => same sequence\n",
      "Line 16: AT3G02870 AT3G02870 => same sequence\n",
      "Line 17: AT3G03260 AT3G03260 => same sequence\n",
      "Line 18: AT4G14790 AT4G14790 => same sequence\n",
      "Line 19: AT4G15210 AT4G15210 => same sequence\n",
      "Line 20: AT4G15560 AT4G15560 => same sequence\n",
      "Line 21: AT4G15570 AT4G15570 => same sequence\n",
      "Line 22: AT4G15802 AT4G15802 => same sequence\n",
      "Line 23: AT4G15880 AT4G15880 => same sequence\n",
      "Line 24: AT4G16420 AT4G16420 => same sequence\n",
      "Line 25: AT4G16480 AT4G16480 => same sequence\n",
      "Line 26: AT5G10480 AT5G10480 => same sequence\n",
      "Line 27: AT5G10510 AT5G10510 => same sequence\n",
      "Line 28: AT5G11110 AT5G11110 => same sequence\n",
      "Line 29: AT5G11260 AT5G11260 => same sequence\n",
      "Line 30: AT5G11510 AT5G11510 => same sequence\n",
      "Line 31: AT5G12200 AT5G12200 => same sequence\n",
      "Line 32: AT5G13290 AT5G13290 => same sequence\n",
      "\n",
      "Files in the same sequence.\n"
     ]
    }
   ],
   "source": [
    "germ = open('Germplasm.tsv' ,'r')\n",
    "locus = open('LocusGene.tsv' ,'r')\n",
    "\n",
    "\n",
    "def read_AGI (file, sep):\n",
    "    try:\n",
    "        agiCode = [line.rstrip().split(sep)[0] for line in file.readlines()]\n",
    "        return agiCode\n",
    "    except AttributeError:\n",
    "        print ('AttributeError: Make sure input is a non-empty file')\n",
    "    \n",
    "\n",
    "\n",
    "def check_seq (file1, file2, sep1, sep2):\n",
    "    separator = ('\\t','\\n','\\r')\n",
    "    try:\n",
    "        if sep1 not in separator or sep2 not in separator:\n",
    "            return print('Wrong separator, try again.')\n",
    "        \n",
    "        agiCode1 = read_AGI(file1, sep1)\n",
    "        agiCode2 = read_AGI(file2, sep2)\n",
    "        length = max(len(agiCode1), len(agiCode2))\n",
    "        \n",
    "        try:\n",
    "            for i in range(0,length): # 1 deletes header\n",
    "                    if agiCode1[i] == agiCode2[i]:                \n",
    "                        print(\"Line {}: {} {} => same sequence\".format(i,agiCode1[i],agiCode2[i]))\n",
    "                    else:\n",
    "                        print(\"Line {}: {} {} => not in the same sequence\".format(i,agiCode1[i],agiCode2[i]))\n",
    "            if agiCode1 == agiCode2:\n",
    "                return print(\"\\nFiles in the same sequence.\")\n",
    "        except:  \n",
    "            return print(\"\\nFiles not in the same sequence.\")\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "\n",
    "check_seq(germ,locus,'\\t','\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Connection to mySQL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sql extension is already loaded. To reload it, use:\n",
      "  %reload_ext sql\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "10 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 243,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql drop database db1;\n",
    "%sql create database db1;\n",
    "%sql show databases;\n",
    "%sql use db1;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Creation of locus and germplasm tables\n",
    "\n",
    " Linked in a 1:1 relationship by the AGI Locus Code where this code is the primary key in the locus table and a foreign key in the germplasm table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "%sql CREATE TABLE LocusGene(Locus VARCHAR(30) NOT NULL, Gene VARCHAR(30) NOT NULL, ProteinLength INT NOT NULL, PRIMARY KEY(Locus));\n",
    "\n",
    "%sql CREATE TABLE Germplasm(id INT NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(30) NOT NULL, germplasm VARCHAR(30) NOT NULL, phenotype VARCHAR(400) NOT NULL, pubmed INT NOT NULL, CONSTRAINT fk_locusCd FOREIGN KEY (Locus) REFERENCES LocusGene(Locus));\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql describe LocusGene;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql describe germplasm;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AT1G01040,DCL1,...added to database\n",
      "AT1G01060,LHY,...added to database\n",
      "AT1G01140,CIPK9,...added to database\n",
      "AT1G01220,FKGP,...added to database\n",
      "AT2G03720,MRH6,...added to database\n",
      "AT2G03800,GEK1,...added to database\n",
      "AT2G04240,XERICO,...added to database\n",
      "AT2G05210,POT1A,...added to database\n",
      "AT3G02130,RPK2,...added to database\n",
      "AT3G02140,TMAC2,...added to database\n",
      "AT3G02230,RGP1,...added to database\n",
      "AT3G02260,BIG,...added to database\n",
      "AT3G02310,SEP2,...added to database\n",
      "AT3G02680,NBS1,...added to database\n",
      "AT3G02850,SKOR,...added to database\n",
      "AT3G02870,VTC4,...added to database\n",
      "AT3G03260,HDG8,...added to database\n",
      "AT4G14790,SUV3,...added to database\n",
      "AT4G15210,BAM5,...added to database\n",
      "AT4G15560,DXS,...added to database\n",
      "AT4G15570,MAA3,...added to database\n",
      "AT4G15802,HSBP,...added to database\n",
      "AT4G15880,ESD4,...added to database\n",
      "AT4G16420,ADA2B,...added to database\n",
      "AT4G16480,INT4,...added to database\n",
      "AT5G10480,PAS2,...added to database\n",
      "AT5G10510,AIL6,...added to database\n",
      "AT5G11110,SPS2,...added to database\n",
      "AT5G11260,HY5,...added to database\n",
      "AT5G11510,MYB3R-4,...added to database\n",
      "AT5G12200,PYD2,...added to database\n",
      "AT5G13290,CRN,...added to database\n",
      " \n",
      "AT1G01040,CS3828,...added to database\n",
      "AT1G01060,lhy-101,...added to database\n",
      "AT1G01140,SALK_058629,...added to database\n",
      "AT1G01220,SALK_012400C,...added to database\n",
      "AT2G03720,SALK_042433,...added to database\n",
      "AT2G03800,gek1-1,...added to database\n",
      "AT2G04240,xerico,...added to database\n",
      "AT2G05210,pot1-1,...added to database\n",
      "AT3G02130,rpk2-2,...added to database\n",
      "AT3G02140,afp4-1,...added to database\n",
      "AT3G02230,rgp1-1,...added to database\n",
      "AT3G02260,tir3-1 RGLG1:rglg1 rglg2,...added to database\n",
      "AT3G02310,sep2-1,...added to database\n",
      "AT3G02680,atnbs1-1,...added to database\n",
      "AT3G02850,CS3816,...added to database\n",
      "AT3G02870,vtc4-1,...added to database\n",
      "AT3G03260,hdg8-1,...added to database\n",
      "AT4G14790,pdd17,...added to database\n",
      "AT4G15210,bmy1-2,...added to database\n",
      "AT4G15560,cla1-1,...added to database\n",
      "AT4G15570,maa3,...added to database\n",
      "AT4G15802,Athspb-2,...added to database\n",
      "AT4G15880,esd4-2,...added to database\n",
      "AT4G16420,prz1-1,...added to database\n",
      "AT4G16480,atint4-2,...added to database\n",
      "AT5G10480,pas2-3,...added to database\n",
      "AT5G10510,plt3-1,...added to database\n",
      "AT5G11110,kns2,...added to database\n",
      "AT5G11260,hy5-101,...added to database\n",
      "AT5G11510,myb3r4-1,...added to database\n",
      "AT5G12200,pyd2-2,...added to database\n",
      "AT5G13290,crn-1,...added to database\n",
      " \n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors \n",
    "import re\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "#connection.autocommit = False\n",
    "\n",
    "def insertValues (db, tname, col1, col2, col3, *args):\n",
    "    try:\n",
    "        with db.cursor() as cursor:\n",
    "            \n",
    "            sql = \"SELECT Locus FROM {table} WHERE Locus = %s\"\n",
    "            cursor.execute(sql.format(table=tname), (col1))\n",
    "            data = cursor.fetchall()\n",
    "            if len(data)!=0:\n",
    "                print('AGI Locus Code \"{}\" already in {}'.format(col1,tname)) \n",
    "            else:\n",
    "                \n",
    "                if args == ():\n",
    "                    sql = \"INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES (%s,%s,%s)\"\n",
    "                    cursor.execute(sql, (col1,col2,col3))\n",
    "    \n",
    "                else:\n",
    "                    sql = \"INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES (%s,%s,%s,%s)\"\n",
    "                    cursor.execute(sql, (col1,col2,col3,args))\n",
    "                \n",
    "                print('{},{},...added to database'.format(col1,col2))\n",
    "                db.commit()\n",
    "                \n",
    "                \n",
    "    except:\n",
    "        print(\"Failed to insert values {},{},...\".format(col1,col2))\n",
    "        \n",
    "    finally:\n",
    "        cursor.close()\n",
    "    \n",
    "def readLines (db,file,tname, sep):\n",
    "    header = [line.rstrip().split(sep) for line in file.readlines(1)]\n",
    "    try:\n",
    "        for line in file.readlines():\n",
    "            line = line.rstrip()\n",
    "            if header == [['Locus', 'Gene', 'ProteinLength']]:\n",
    "                Locus, Gene, ProteinLength = line.split(sep)\n",
    "                insertValues(db,tname, Locus, Gene, ProteinLength)\n",
    "            else:\n",
    "                Locus, germplasm, phenotype, pubmed = line.split(sep)\n",
    "                insertValues(db,tname, Locus, germplasm, phenotype, pubmed)\n",
    "    finally:\n",
    "        print(\" \")        \n",
    "\n",
    "\n",
    "\n",
    "LocusGene = open(\"LocusGene.tsv\", \"r\")\n",
    "readLines(connection, LocusGene, 'LocusGene', '\\t')\n",
    "\n",
    "Germplasm = open(\"Germplasm.tsv\", \"r\")\n",
    "readLines(connection, Germplasm, 'Germplasm', '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Nº queries added: 4\n",
      "\n",
      "File closed\n",
      "\n",
      "Database closed\n"
     ]
    }
   ],
   "source": [
    "file = open('report.txt', 'w')\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='db1',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "#connection.autocommit = False\n",
    "\n",
    "def createReport (db, queries, sep):\n",
    "    \"\"\"\n",
    "    Funcionction that write MySQL queries in an empty file. \n",
    "    db = name of connection.\n",
    "    queries = List of queries with MySQL format.\n",
    "    sep = tab or comma\n",
    "    \"\"\"\n",
    "    count=0\n",
    "    if sep not in ('\\t', ','):\n",
    "        print (\"Separator not valid, try again with '\\\\t' or ','.\")\n",
    "    else:\n",
    "        try:\n",
    "            if queries != []:\n",
    "                for query in queries:\n",
    "                    file.write('\\n>Query nº{} :\"{}\"\\n'.format(queries.index(query)+1,query))\n",
    "                    try:\n",
    "                        with db.cursor() as cursor:\n",
    "                            cursor.execute(query)\n",
    "                            results = cursor.fetchall()\n",
    "                            if len(results)==0:\n",
    "                                print('No data to report') \n",
    "                            else:\n",
    "                                count+=1\n",
    "                                header =(sep.join(list(results[0].keys())) + '\\n')\n",
    "                                file.write(header)\n",
    "                                for result in results:\n",
    "                                    newLine = [str(value) for value in result.values()]\n",
    "                                    file.write(str(sep.join(newLine)+ '\\n'))           \n",
    "                    except:\n",
    "                        print(\"Failed to write query\")\n",
    "            else:\n",
    "                print(\"Error: Varliable 'queries' is empty, please try again.\")\n",
    "        except:\n",
    "            print('Something went wrong: Please check input variables')\n",
    "\n",
    "        finally:\n",
    "            print(\"\\nNº queries added:\" , count)\n",
    "            file.close()\n",
    "            print('\\nFile closed')\n",
    "            db.cursor().close()\n",
    "            print('\\nDatabase closed')\n",
    "\n",
    "        \n",
    "\n",
    "    \n",
    "query1=\"SELECT * FROM LocusGene JOIN Germplasm ON LocusGene.Locus = Germplasm.Locus\"\n",
    "query2=\"SELECT * FROM LocusGene JOIN Germplasm ON LocusGene.Locus = Germplasm.Locus WHERE Gene in ('SKOR', 'MAA3')\"\n",
    "query3=\"SELECT SUBSTRING(Locus,1,4),COUNT(Locus) FROM LocusGene GROUP BY SUBSTRING(Locus,1,4);\"\n",
    "query4=\"SELECT SUBSTRING(Locus,1,4),AVG(ProteinLength) FROM LocusGene GROUP BY SUBSTRING(Locus,1,4);\"\n",
    "\n",
    "queries = [query1, query2, query3, query4]\n",
    "\n",
    "createReport(connection,queries, '\\t') \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check if reports have been added to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      ">Query nº1 :\"SELECT * FROM LocusGene JOIN Germplasm ON LocusGene.Locus = Germplasm.Locus\"\n",
      "Locus\tGene\tProteinLength\tid\tGermplasm.Locus\tgermplasm\tphenotype\tpubmed\n",
      "AT1G01040\tDCL1\t332\t1\tAT1G01040\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\n",
      "AT1G01060\tLHY\t290\t2\tAT1G01060\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\n",
      "AT1G01140\tCIPK9\t223\t3\tAT1G01140\tSALK_058629\thypersensitive to low potassium media\t17486125\n",
      "AT1G01220\tFKGP\t190\t4\tAT1G01220\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\n",
      "AT2G03720\tMRH6\t189\t5\tAT2G03720\tSALK_042433\tMultiple straight hairs\t16367956\n",
      "AT2G03800\tGEK1\t196\t6\tAT2G03800\tgek1-1\tEthanol hypersensitivity.\t15215505\n",
      "AT2G04240\tXERICO\t256\t7\tAT2G04240\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\n",
      "AT2G05210\tPOT1A\t221\t8\tAT2G05210\tpot1-1\tNo visible phenotype.\t17627276\n",
      "AT3G02130\tRPK2\t284\t9\tAT3G02130\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\n",
      "AT3G02140\tTMAC2\t300\t10\tAT3G02140\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\n",
      "AT3G02230\tRGP1\t301\t11\tAT3G02230\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\n",
      "AT3G02260\tBIG\t279\t12\tAT3G02260\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\n",
      "AT3G02310\tSEP2\t175\t13\tAT3G02310\tsep2-1\tNon-described subtle phenotype.\t10821278\n",
      "AT3G02680\tNBS1\t190\t14\tAT3G02680\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\n",
      "AT3G02850\tSKOR\t234\t15\tAT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\n",
      "AT3G02870\tVTC4\t311\t16\tAT3G02870\tvtc4-1\tascorbate deficient\t16595667\n",
      "AT3G03260\tHDG8\t194\t17\tAT3G03260\thdg8-1\tNo visible phenotype.\t16778018\n",
      "AT4G14790\tSUV3\t312\t18\tAT4G14790\tpdd17\tDefective pollen development.\t19237690\n",
      "AT4G15210\tBAM5\t313\t19\tAT4G15210\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\n",
      "AT4G15560\tDXS\t219\t20\tAT4G15560\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\n",
      "AT4G15570\tMAA3\t294\t21\tAT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\n",
      "AT4G15802\tHSBP\t254\t22\tAT4G15802\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\n",
      "AT4G15880\tESD4\t265\t23\tAT4G15880\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\n",
      "AT4G16420\tADA2B\t279\t24\tAT4G16420\tprz1-1\tAltered response to auxin and cytokinin\t12747832\n",
      "AT4G16480\tINT4\t284\t25\tAT4G16480\tatint4-2\tNo visible phenotype.\t16603666\n",
      "AT5G10480\tPAS2\t301\t26\tAT5G10480\tpas2-3\tSegregates 25% embryo lethal.\t18799749\n",
      "AT5G10510\tAIL6\t310\t27\tAT5G10510\tplt3-1\tShort roots and shortened root meristem.\t17960244\n",
      "AT5G11110\tSPS2\t232\t28\tAT5G11110\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\n",
      "AT5G11260\tHY5\t221\t29\tAT5G11260\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\n",
      "AT5G11510\tMYB3R-4\t336\t30\tAT5G11510\tmyb3r4-1\tNo visible phenotype.\t17287251\n",
      "AT5G12200\tPYD2\t310\t31\tAT5G12200\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\n",
      "AT5G13290\tCRN\t189\t32\tAT5G13290\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\n",
      "\n",
      ">Query nº2 :\"SELECT * FROM LocusGene JOIN Germplasm ON LocusGene.Locus = Germplasm.Locus WHERE Gene in ('SKOR', 'MAA3')\"\n",
      "Locus\tGene\tProteinLength\tid\tGermplasm.Locus\tgermplasm\tphenotype\tpubmed\n",
      "AT3G02850\tSKOR\t234\t15\tAT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\n",
      "AT4G15570\tMAA3\t294\t21\tAT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\n",
      "\n",
      ">Query nº3 :\"SELECT SUBSTRING(Locus,1,4),COUNT(Locus) FROM LocusGene GROUP BY SUBSTRING(Locus,1,4);\"\n",
      "SUBSTRING(Locus,1,4)\tCOUNT(Locus)\n",
      "AT1G\t4\n",
      "AT2G\t4\n",
      "AT3G\t9\n",
      "AT4G\t8\n",
      "AT5G\t7\n",
      "\n",
      ">Query nº4 :\"SELECT SUBSTRING(Locus,1,4),AVG(ProteinLength) FROM LocusGene GROUP BY SUBSTRING(Locus,1,4);\"\n",
      "SUBSTRING(Locus,1,4)\tAVG(ProteinLength)\n",
      "AT1G\t258.7500\n",
      "AT2G\t215.5000\n",
      "AT3G\t252.0000\n",
      "AT4G\t277.5000\n",
      "AT5G\t271.2857\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "check = open('report.txt', 'r')\n",
    "print(check.read())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "</pre>\n",
    "## Don't forget to commit and push your answers before you leave!\n",
    "\n",
    "It was wonderful to have you in my class!  I hope to see you again soon!\n",
    "\n",
    "Good luck with your careers!!\n",
    "\n",
    "Mark"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
